File: Source/Game/Entities/NamedWolves.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class NamedWolves
{
    public static List<Wolf> DNTNamedWolves { get; set; } = new List<Wolf>();
    public static Wolf ExplodingWolf;
    public static Wolf StanWolf;
    public const string STAN_NAME = $"{Colors.COLOR_PURPLE}Stan|r";
    private static timer ExplodingTexttagTimer = timer.Create();
    private static timer ExplodingWolfRevive = timer.Create();
    private static string BLOOD_EFFECT_PATH = "war3mapImported\\Bloodstrike.mdx";

    /// <summary>
    /// Creates the named Wolves Marco and Stan. Only works in Standard mode.
    /// </summary>
    public static void CreateNamedWolves()
    {
        if (Gamemode.CurrentGameMode != "Standard") return;
        DNTNamedWolves.Clear();
        CreateStanWolf();
        CreateExplodingWolf();
    }

    private static void CreateExplodingWolf()
    {
        var index = GetRandomInt(0, RegionList.WolfRegions.Length - 1);
        ExplodingWolf = new Wolf(index);
        ExplodingWolf.Texttag.SetPermanent(true);
        ExplodingTexttagTimer.Start(0.03f, true, () => ExplodingWolf.Texttag.SetPosition(ExplodingWolf.Unit.X, ExplodingWolf.Unit.Y, 0.015f));
        ExplodingWolfDesc();
    }

    private static void CreateStanWolf()
    {
        var index = GetRandomInt(0, RegionList.WolfRegions.Length - 1);
        StanWolf = new Wolf(index);

        StanWolf.WanderTimer.Pause();
        StanWolf.WanderTimer = null;
        StanWolf.Unit.SetVertexColor(235, 115, 255);
        StanWolf.Unit.Name = STAN_NAME;

        StanWolf.Texttag ??= texttag.Create();
        StanWolf.Texttag.SetText(StanWolf.Unit.Name, 0.015f);
        StanWolf.Texttag.SetPermanent(true);
        StanWolf.IsPaused = true;

        StanWolf.Unit.IsInvulnerable = false;
        BurntMeat.RegisterDeathTrigger();

        Utility.SimpleTimer(0.03f, () => StanWolf.Texttag.SetPosition(StanWolf.Unit.X, StanWolf.Unit.Y, 0.015f));
        DNTNamedWolves.Add(StanWolf);
    }

    public static void KillExplodingWolf()
    {
        try
        {
            if (ExplodingWolf.IsReviving) return;
            ExplodingWolf.Unit.Kill();
            ExplodingWolf.IsReviving = true;
            ExplodingWolf.OVERHEAD_EFFECT_PATH = "";
            ExplodingWolf.Texttag.SetText("", 0.015f);
            Utility.CreateEffectAndDispose(BLOOD_EFFECT_PATH, ExplodingWolf.Unit, "origin");
            ExplodingWolfRevive.Start(25.0f, false, () =>
            {
                DNTNamedWolves.Remove(ExplodingWolf);
                ExplodingWolf.IsReviving = false;
                ExplodingWolf.Unit?.Dispose();
                Globals.ALL_WOLVES.Remove(ExplodingWolf.Unit);
                ExplodingWolf.Unit = unit.Create(ExplodingWolf.Unit.Owner, Wolf.WOLF_MODEL, ExplodingWolf.Unit.X, ExplodingWolf.Unit.Y, 360);
                Globals.ALL_WOLVES.Add(ExplodingWolf.Unit, ExplodingWolf);
                ExplodingWolfDesc();
            });
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in KillExplodingWolf {e.Message}");
        }
    }

    private static void ExplodingTexttag()
    {
        ExplodingWolf.Texttag ??= texttag.Create();
        ExplodingWolf.Texttag.SetText(ExplodingWolf.Unit.Name, 0.015f);
    }

    private static void ExplodingWolfDesc()
    {
        try
        {
            Utility.MakeUnitLocust(ExplodingWolf.Unit);
            var randomPlayer = GetRandomPlayerFromLobby();
            SetRandomVertexColor(ExplodingWolf.Unit, randomPlayer.Id);
            ExplodingWolf.Unit.Name = Utility.FormattedColorPlayerName(randomPlayer);
            ExplodingTexttag();
            ExplodingWolf.OVERHEAD_EFFECT_PATH = Wolf.DEFAULT_OVERHEAD_EFFECT;
            DNTNamedWolves.Add(ExplodingWolf);
        }
        catch (Exception e)
        {
            Logger.Warning($"Error in ExplodingWolfDesc {e.Message}");
        }
    }

    public static bool ExplodingWolfCollision(unit unit, Kitty k)
    {
        if (Gamemode.CurrentGameMode != "Standard") return false;
        if (unit != ExplodingWolf.Unit) return false;
        Utility.GiveGoldFloatingText(25, k.Unit);
        BurntMeat.FlamesDropChance(k);
        KillExplodingWolf();
        return true;
    }

    private static player GetRandomPlayerFromLobby()
    {
        return Globals.ALL_PLAYERS[GetRandomInt(0, Globals.ALL_PLAYERS.Count - 1)];
    }

    private static void SetRandomVertexColor(unit u, int playerID)
    {
        Colors.SetUnitToVertexColor(u, playerID);
    }
}

File: src/Game/Entities/NamedWolves.ts
import { Logger } from 'src/Events/Logger/Logger'
import { Gamemode } from 'src/Gamemodes/Gamemode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { RegionList } from 'src/Global/RegionList'
import { BurntMeat } from 'src/Rewards/EasterEggs/BurntMeat'
import { Colors } from 'src/Utility/Colors/Colors'
import { ErrorHandler } from 'src/Utility/ErrorHandler'
import { Utility } from 'src/Utility/Utility'
import { MapPlayer, TextTag, Timer, Unit } from 'w3ts'
import { Kitty } from './Kitty/Kitty'
import { Wolf } from './Wolf'

export class NamedWolves {
    public static DNTNamedWolves: Wolf[] = []
    public static ExplodingWolf: Wolf
    public static StanWolf: Wolf
    public STAN_NAME: string = '{Colors.COLOR_PURPLE}Stan|r'
    private static ExplodingTexttagTimer = Timer.create()
    private static ExplodingWolfRevive = Timer.create()
    private static BLOOD_EFFECT_PATH: string = 'war3mapImported\\Bloodstrike.mdx'

    /// <summary>
    /// Creates the named Wolves Marco and Stan. Only works in Standard mode.
    /// </summary>
    public static CreateNamedWolves() {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return
        NamedWolves.DNTNamedWolves.clear()
        NamedWolves.CreateStanWolf()
        NamedWolves.CreateExplodingWolf()
    }

    private static CreateExplodingWolf() {
        let index = GetRandomInt(0, RegionList.WolfRegions.length - 1)
        NamedWolves.ExplodingWolf = new Wolf(index)
        NamedWolves.ExplodingWolf.Texttag.setPermanent(true)
        NamedWolves.ExplodingTexttagTimer.start(0.03, true, NamedWolves.UpdateTextTag)
        NamedWolves.ExplodingWolfDesc()
    }

    private static UpdateTextTag() {
        NamedWolves.ExplodingWolf?.Texttag?.setPos(
            NamedWolves.ExplodingWolf.Unit.x,
            NamedWolves.ExplodingWolf.unit.y,
            0.015
        )
    }

    private static CreateStanWolf() {
        let index = GetRandomInt(0, RegionList.WolfRegions.length - 1)
        NamedWolves.StanWolf = new Wolf(index)

        NamedWolves.StanWolf.PauseSelf(true)
        NamedWolves.StanWolf.Unit.setVertexColor(235, 115, 255, 255)
        NamedWolves.StanWolf.Unit.name = STAN_NAME
        NamedWolves.StanWolf.OVERHEAD_EFFECT_PATH = ''

        NamedWolves.StanWolf.Texttag ??= TextTag.create()!
        NamedWolves.StanWolf.Texttag.setText(NamedWolves.StanWolf.Unit.name, 0.015)
        NamedWolves.StanWolf.Texttag.setPermanent(true)
        NamedWolves.StanWolf.paused = true
        NamedWolves.StanWolf.Unit.invulnerable = false

        Utility.SimpleTimer(0.5, () => (NamedWolves.StanWolf.Unit.invulnerable = false))
        BurntMeat.RegisterDeathTrigger()

        Utility.SimpleTimer(0.5, () =>
            NamedWolves.StanWolf.Texttag.setPos(NamedWolves.StanWolf.Unit.x, NamedWolves.StanWolf.unit.y, 0.015)
        )
        NamedWolves.DNTNamedWolves.push(NamedWolves.StanWolf)
    }

    public static KillExplodingWolf() {
        try {
            if (NamedWolves.ExplodingWolf.IsReviving) return
            NamedWolves.ExplodingWolf.Unit.Kill()
            NamedWolves.ExplodingWolf.IsReviving = true
            NamedWolves.ExplodingWolf.OVERHEAD_EFFECT_PATH = ''
            NamedWolves.ExplodingWolf.Texttag.setText('', 0.015)
            Utility.CreateEffectAndDisposeAttach(
                NamedWolves.BLOOD_EFFECT_PATH,
                NamedWolves.ExplodingWolf.Unit,
                'origin'
            )
            NamedWolves.ExplodingWolfRevive.start(
                25.0,
                false,
                ErrorHandler.Wrap(() => {
                    if (NamedWolves.ExplodingWolf.Unit == null) return
                    NamedWolves.DNTNamedWolves.Remove(NamedWolves.ExplodingWolf)
                    NamedWolves.ExplodingWolf.IsReviving = false
                    NamedWolves.ExplodingWolf.Unit?.dispose()
                    Globals.ALL_WOLVES.Remove(NamedWolves.ExplodingWolf.Unit)
                    NamedWolves.ExplodingWolf.Unit = Unit.create(
                        NamedWolves.ExplodingWolf.Unit.owner,
                        Wolf.WOLF_MODEL,
                        NamedWolves.ExplodingWolf.Unit.x,
                        NamedWolves.ExplodingWolf.unit.y,
                        360
                    )
                    Globals.ALL_WOLVES.push(NamedWolves.ExplodingWolf.Unit, NamedWolves.ExplodingWolf)
                    NamedWolves.ExplodingWolfDesc()
                })
            )
        } catch (e: any) {
            Logger.Warning('Error in KillExplodingWolf {e.Message}')
        }
    }

    private static ExplodingTexttag() {
        NamedWolves.ExplodingWolf.Texttag ??= TextTag.create()!
        NamedWolves.ExplodingWolf.Texttag.setText(NamedWolves.ExplodingWolf.Unit.name, 0.015)
    }

    private static ExplodingWolfDesc() {
        try {
            Utility.MakeUnitLocust(NamedWolves.ExplodingWolf.Unit)
            let randomPlayer = NamedWolves.GetRandomPlayerFromLobby()
            NamedWolves.SetRandomVertexColor(NamedWolves.ExplodingWolf.Unit, randomPlayer.id)
            NamedWolves.ExplodingWolf.Unit.name = Utility.FormattedColorPlayerName(randomPlayer)
            NamedWolves.ExplodingTexttag()
            NamedWolves.ExplodingWolf.OVERHEAD_EFFECT_PATH = Wolf.DEFAULT_OVERHEAD_EFFECT
            NamedWolves.DNTNamedWolves.push(NamedWolves.ExplodingWolf)
        } catch (e: any) {
            Logger.Warning('Error in ExplodingWolfDesc {e.Message} {e.StackTrace}')
        }
    }

    public static ExplodingWolfCollision(unit: Unit, k: Kitty, shadowKitty: boolean = false) {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return false
        if (unit != NamedWolves.ExplodingWolf.Unit) return false
        if (!shadowKitty) Utility.GiveGoldFloatingText(25, k.Unit)
        else Utility.GiveGoldFloatingText(25, k.ShadowKitty.Unit)
        BurntMeat.FlamesDropChance(k)
        NamedWolves.KillExplodingWolf()
        return true
    }

    private static GetRandomPlayerFromLobby(): MapPlayer {
        return Globals.ALL_PLAYERS[GetRandomInt(0, Globals.ALL_PLAYERS.length - 1)]
    }

    private static SetRandomVertexColor(u: Unit, playerID: number) {
        Colors.SetUnitToVertexColor(u, playerID)
    }
}
